packages/lib/services/searchengine/SearchEngine.ts

```diff
@@ -554,6 +554,94 @@ export default class SearchEngine {
 		return SearchEngine.SEARCH_TYPE_FTS;
 	}
 
+	makeSearchTree(str: string) {
+		// (tag:A (any:1 tag:B tag:C)) -> [ "tag:A", [ "any:1", "tag:B", "tag:C"]]
+		let i = 0;
+		function main(): any {
+			const arr = [];
+			let startIndex = i;
+			function addWord() {
+				if (i - 1 > startIndex) {
+					arr.push(str.slice(startIndex, i - 1));
+				}
+			}
+			while (i < str.length) {
+				switch (str[i++]) {
+				case ' ':
+					addWord();
+					startIndex = i;
+					continue;
+				case '(':
+					arr.push(main());
+					startIndex = i;
+					continue;
+				case ')':
+					addWord();
+					return arr;
+				}
+			}
+			addWord();
+			return arr;
+		}
+		return main();
+	}
+
+
+
+	async makeQuery(parseTree: any) {
+
+		const getQueryWithParms = async (searchString: string) => {
+			const parsedQuery = await this.parseQuery(searchString);
+			const { query, params } = queryBuilder(parsedQuery.allTerms);
+			return { query, params };
+		};
+
+		const intersection = (queries: any) => {
+			const result = queries.reduce((acc: any, curr: any, idx: number) => {
+				return {
+					query: `SELECT * FROM (${acc.query}) t${idx - 1} INNER JOIN (${curr.query}) t${idx} ON t${idx - 1}.id=t${idx}.id`,
+					params: acc.params.concat(curr.params),
+				};
+			});
+			return result;
+		};
+
+		const union = (queries: any) => {
+			const query = `SELECT * FROM (${queries.map((x: any) => x.query).join(' UNION ')})`;
+			const params = queries.map((x: any)=>x.params).flat();
+			return { query, params };
+		};
```

Comment:
Doing some additional testing made it clear these union and concatenation functions are not enough. They introduce duplicate results, so I'll need to rework these SQL queries.

I'm also busy finishing up my thesis work, so this is going to take more time to get done. Sorry about that. I'll let you know when I've made some progress.

Reply:
Sounds good, and best of luck with your thesis.